File: client/python/src/hiver/sandbox.py
from __future__ import annotations

import asyncio
import json
import uuid
from typing import AsyncGenerator, Callable, Coroutine, Any, AsyncIterable, Optional, Union
from urllib.parse import quote

import httpx

from .schemas import (
    ApplyResult,
    SandboxConfig,
    SandboxEvent,
    SandboxEventAdapter,
    SandboxInfo,
    SandboxRef,
    Snapshot,
    SnapshotResult,
)
from .sse import parse_sse

_FETCH_TIMEOUT = httpx.Timeout(connect=3.0, read=3.0, write=3.0, pool=3.0)


class SandboxError(Exception):
    """Raised when a client operation against a sandbox or the controller fails."""

    def __init__(
        self,
        operation: str,
        status: int,
        message: str,
        body: Optional[dict] = None,
    ) -> None:
        super().__init__(f"{operation}: {message}")
        self.status = status
        """HTTP status from the failed response, or ``0`` if the request never reached the server."""
        self.operation = operation
        """The client operation that failed (e.g. ``"apply_config"``)."""
        self.body = body
        """Structured error payload from the server, when one was returned."""


def _to_error(res: httpx.Response, operation: str) -> SandboxError:
    text = res.text
    body: Optional[dict] = None
    try:
        parsed = json.loads(text)
        if isinstance(parsed, dict) and "error" in parsed:
            body = parsed
    except Exception:
        pass
    message = (body or {}).get("error") or text or str(res.status_code)
    return SandboxError(operation, res.status_code, message, body)


class ExecProcess:
    """
    Handle to an interactive command started with :meth:`Sandbox.exec_stream`.
    Stream output via :attr:`pipes`, send input via :meth:`write_stdin`, and
    await the result via :attr:`exit_code`.
    """

    id: str
    """Unique id for this exec invocation."""
    pipes: AsyncIterable[dict[str, str]]
    """Async iterable of output chunks emitted as the process runs — each is ``{"stdout": ...}`` or ``{"stderr": ...}``."""
    exit_code: "asyncio.Future[int]"
    """Resolves with the process exit code once it finishes."""

    def __init__(
        self,
        id: str,
        pipes: AsyncIterable[dict[str, str]],
        exit_code: asyncio.Future[int],
        write_stdin_fn: Callable[[str], Coroutine[Any, Any, None]],
    ) -> None:
        self.id = id
        self.pipes = pipes
        self.exit_code = exit_code
        self._write_stdin_fn = write_stdin_fn

    async def write_stdin(self, data: str) -> None:
        """Send `data` to the process's stdin."""
        await self._write_stdin_fn(data)


class Sandbox:
    """
    Handle to a provisioned sandbox. Returned by :func:`get_or_create_sandbox`;
    not constructed directly by callers.
    """

    id: str
    """Server-assigned unique identifier (uuid)."""
    key: str
    """Caller-chosen key the sandbox was provisioned under; routes requests."""
    api_server_url: str
    """Base URL of the per-sandbox API server."""

    def __init__(
        self,
        ref: SandboxRef,
        gateway_url: str,
        client: Optional[httpx.AsyncClient] = None,
    ) -> None:
        self.id = ref.id
        self.key = ref.key
        self.api_server_url = f"{gateway_url.rstrip('/')}/sandbox/{ref.id}"
        self._owns_client = client is None
        self._client = client or httpx.AsyncClient(timeout=_FETCH_TIMEOUT)

    def proxy_url(self, port: Union[int, str]) -> str:
        """
        Base proxy URL for reaching a port inside the sandbox. Ends with a
        trailing slash, so it reaches the port's root as-is; append a path to
        reach an endpoint, e.g. ``sandbox.proxy_url(8080) + "health"``.
        """
        return f"{self.api_server_url}/v1/{self.key}/proxy/{port}/"

    async def aclose(self) -> None:
        """Close the underlying HTTP client if this sandbox owns it."""
        if self._owns_client:
            await self._client.aclose()

    async def __aenter__(self) -> "Sandbox":
        return self

    async def __aexit__(self, *_: object) -> None:
        await self.aclose()

    async def ping(self) -> None:
        """Keep the sandbox alive by resetting its TTL countdown."""
        res = await self._client.get(f"{self.api_server_url}/v1/{self.key}/ping")
        if not res.is_success:
            raise _to_error(res, "ping")

    async def shutdown(self) -> None:
        """Tear this sandbox down via ``DELETE /v1/<key>``, cancelling its lifecycle."""
        res = await self._client.delete(f"{self.api_server_url}/v1/{self.key}")
        if not res.is_success:
            raise _to_error(res, "shutdown")

    async def get_ports(self) -> list[int]:
        """
        List the network ports the sandbox exposes. Reach each one via
        :meth:`proxy_url`.
        """
        res = await self._client.get(f"{self.api_server_url}/v1/{self.key}/ports")
        if not res.is_success:
            raise _to_error(res, "get_ports")
        return res.json()

    async def get_info(self) -> SandboxInfo:
        """
        Read internal runtime info about the sandbox — currently the isolation
        mechanism in use, which is selected automatically from the image (a
        microvm image ships a guest root filesystem) rather than configured.
        """
        res = await self._client.get(f"{self.api_server_url}/v1/{self.key}/info")
        if not res.is_success:
            raise _to_error(res, "get_info")
        return SandboxInfo.model_validate(res.json())

    async def get_config(self) -> SandboxConfig:
        """Read the current SandboxConfig."""
        res = await self._client.get(f"{self.api_server_url}/v1/{self.key}/config")
        if not res.is_success:
            raise _to_error(res, "get_config")
        return SandboxConfig.model_validate(res.json())

    async def apply_config(self, config: SandboxConfig) -> ApplyResult:
        """
        Apply a new SandboxConfig. The change is all-or-nothing: the returned
        ApplyResult's ``applied`` field reports whether it was committed or
        rolled back, and ``changes`` details what was added or removed.
        """
        validated = SandboxConfig.model_validate(config.model_dump(exclude_none=True))
        res = await self._client.put(
            f"{self.api_server_url}/v1/{self.key}/config",
            json=validated.model_dump(exclude_none=True),
        )
        if not res.is_success:
            raise _to_error(res, "apply_config")
        return ApplyResult.model_validate(res.json())

    async def snapshot(self, request: Snapshot) -> SnapshotResult:
        """
        Capture a snapshot of the running sandbox now, without stopping it. The
        request selects which parts to capture: ``vm`` (full microVM state, keyed
        for a later resume; a no-op on container isolation) and/or ``files`` (the
        writable filesystem). Each part is reported independently in the result.
        """
        validated = Snapshot.model_validate(request.model_dump(exclude_none=True))
        res = await self._client.post(
            f"{self.api_server_url}/v1/{self.key}/snapshot",
            json=validated.model_dump(exclude_none=True),
        )
        if not res.is_success:
            raise _to_error(res, "snapshot")
        return SnapshotResult.model_validate(res.json())

    async def exec(
        self,
        command: Union[str, list[str]],
        cwd: Optional[str] = None,
        env: Optional[dict[str, str]] = None,
    ) -> dict[str, object]:
        """
        Run `command` inside the sandbox and return buffered stdout, stderr,
        and exit_code once the process finishes.

        `command` may be a string (passed to a shell via ``sh -c``) or a list of
        strings (executed directly as argv, each element a literal argument with
        no shell, word-splitting, or expansion).

        `env` is merged on top of the sandbox config's environment, overriding
        entries with the same name. When omitted, the sandbox config
        environment is used as-is.
        """
        body: dict[str, object] = {"command": command}
        if cwd is not None:
            body["cwd"] = cwd
        if env is not None:
            body["env"] = env
        res = await self._client.post(
            f"{self.api_server_url}/v1/{self.key}/exec",
            json=body,
        )
        if not res.is_success:
            raise _to_error(res, "exec")
        return res.json()

    async def exec_stream(
        self,
        command: Union[str, list[str]] = "",
        cwd: Optional[str] = None,
        tty: bool = False,
        env: Optional[dict[str, str]] = None,
    ) -> ExecProcess:
        """
        Run `command` inside the sandbox and return an ExecProcess handle for
        interactive use: stream output via ``exec.pipes``, send input via
        ``exec.write_stdin()``, and await the result via ``exec.exit_code``.

        `command` may be a string (passed to a shell via ``sh -c``) or a list of
        strings (executed directly as argv, each element a literal argument with
        no shell, word-splitting, or expansion).

        Pass an empty `command` to attach to the sandbox entrypoint's terminal
        instead of running a new command — this requires the sandbox to have
        been created with ``tty=True``. The stream stays open until the
        entrypoint exits or you disconnect.

        `env` is merged on top of the sandbox config's environment, overriding
        entries with the same name. When omitted, the sandbox config
        environment is used as-is.

        Resolves once the process is ready, so ``write_stdin`` is safe to call.
        """
        exec_id = str(uuid.uuid4())
        stream_url = f"{self.api_server_url}/v1/{self.key}/exec-stream/{exec_id}"
        stdin_url = f"{self.api_server_url}/v1/{self.key}/exec-stream/{exec_id}/stdin"

        body: dict[str, object] = {}
        if command:
            body["command"] = command
        if cwd is not None:
            body["cwd"] = cwd
        if env is not None:
            body["env"] = env
        if tty:
            body["tty"] = tty

        queue: asyncio.Queue[Optional[dict[str, str]]] = asyncio.Queue()
        exit_future: asyncio.Future[int] = asyncio.get_running_loop().create_future()
        ready = asyncio.Event()
        start_error: list[Exception] = []

        client = self._client

        async def _reader() -> None:
            sse_timeout = httpx.Timeout(None, connect=_FETCH_TIMEOUT.connect)
            try:
                async with client.stream(
                    "POST",
                    stream_url,
                    json=body,
                    headers={"accept": "text/event-stream"},
                    timeout=sse_timeout,
                ) as res:
                    if not res.is_success:
                        await res.aread()
                        start_error.append(_to_error(res, "exec_stream"))
                        ready.set()
                        return
                    ready.set()
                    async for frame in parse_sse(res, None):
                        event: dict[str, object] = json.loads(frame.data)
                        etype = event.get("type")
                        if etype == "stdout":
                            await queue.put({"stdout": str(event["text"])})
                        elif etype == "stderr":
                            await queue.put({"stderr": str(event["text"])})
                        elif etype == "exit":
                            if not exit_future.done():
                                exit_future.set_result(int(event["code"]))  # type: ignore[arg-type]
                            await queue.put(None)
            except Exception as exc:
                if not ready.is_set():
                    start_error.append(exc)
                    ready.set()
                if not exit_future.done():
                    exit_future.set_exception(exc)
                await queue.put(None)

        asyncio.create_task(_reader())
        await ready.wait()
        if start_error:
            raise start_error[0]

        async def _pipes() -> AsyncGenerator[dict[str, str], None]:
            while True:
                item = await queue.get()
                if item is None:
                    return
                yield item

        async def _write_stdin(data: str) -> None:
            res = await client.post(stdin_url, json={"data": data})
            if not res.is_success:
                raise _to_error(res, "exec_stream_stdin")

        return ExecProcess(
            id=exec_id,
            pipes=_pipes(),
            exit_code=exit_future,
            write_stdin_fn=_write_stdin,
        )

    async def list_directory(self, path: str) -> list[dict[str, object]]:
        """
        List the immediate children of a directory under a sandbox mount.
        `path` is the agent-visible absolute path (e.g. `/workspace`).
        Returns a list of entries with `name`, `path`, `is_dir`, and `size`.
        """
        res = await self._client.get(
            f"{self.api_server_url}/v1/{self.key}/directories",
            params={"path": path},
        )
        if not res.is_success:
            raise _to_error(res, "list_directory")
        return res.json()["entries"]

    def _file_url(self, path: str) -> str:
        """
        Build the `/file` URL for an agent-visible absolute path, carrying it as
        trailing URL segments (e.g. `/file/workspace/data.csv`). Each segment is
        escaped while the `/` separators are preserved, so a nested path with
        arbitrarily many segments round-trips intact.
        """
        segments = "/".join(quote(seg, safe="") for seg in path.split("/") if seg)
        return f"{self.api_server_url}/v1/{self.key}/file/{segments}"

    async def read_file(self, path: str) -> bytes:
        """
        Download a file from a sandbox mount. `path` is the agent-visible
        absolute path (e.g. `/workspace/data.csv`). Returns the raw bytes.
        """
        res = await self._client.get(self._file_url(path))
        if not res.is_success:
            raise _to_error(res, "read_file")
        return res.content

    async def write_file(
        self,
        path: str,
        content: Union[bytes, str],
    ) -> dict[str, object]:
        """
        Upload `content` as a file to `path`, the agent-visible absolute path
        (e.g. `/workspace/data.csv`), which must resolve beneath one of the
        configured `fs[].mount` paths. Returns the agent-visible path and byte
        count the server reports.
        """
        if isinstance(content, str):
            content = content.encode("utf-8")
        res = await self._client.post(
            self._file_url(path),
            content=content,
            headers={"Content-Type": "application/octet-stream"},
        )
        if not res.is_success:
            raise _to_error(res, "write_file")
        return res.json()

    async def delete_file(self, path: str) -> None:
        """
        Delete a file or empty directory at `path` inside a sandbox mount.
        `path` is the agent-visible absolute path (e.g. `/workspace/data.csv`).
        """
        res = await self._client.delete(self._file_url(path))
        if not res.is_success:
            raise _to_error(res, "delete_file")

    async def get_events_stream(
        self,
        last_event_id: Optional[int] = None,
        abort: Optional[asyncio.Event] = None,
        max_retries: int = 3,
    ) -> AsyncGenerator[SandboxEvent, None]:
        """
        Stream the sandbox's activity events (egress, filesystem, exec, stdio,
        resource usage) as they happen.

        Auto-resumes across transient disconnects without dropping events,
        reconnecting up to ``max_retries`` times. Resume from a known position
        with ``last_event_id``. Stop the stream by setting ``abort`` or
        cancelling the calling task.
        """
        backoff_s = 0.2
        retry = 0

        while abort is None or not abort.is_set():
            if retry > max_retries:
                return
            try:
                async for event in self._open_events_stream(last_event_id, abort):
                    last_event_id = event.id  # type: ignore[union-attr]
                    backoff_s = 0.2
                    yield event
            except asyncio.CancelledError:
                return
            except Exception:
                if abort is not None and abort.is_set():
                    return

            await _sleep(backoff_s, abort)
            backoff_s = min(backoff_s * 2, 30.0)
            retry += 1

    async def _open_events_stream(
        self,
        last_event_id: Optional[int],
        abort: Optional[asyncio.Event],
    ) -> AsyncGenerator[SandboxEvent, None]:
        params: dict[str, str] = {}
        if last_event_id is not None:
            params["lastEventId"] = str(last_event_id)

        sse_timeout = httpx.Timeout(None, connect=_FETCH_TIMEOUT.connect)
        async with self._client.stream(
            "GET",
            f"{self.api_server_url}/v1/{self.key}/events",
            headers={"accept": "text/event-stream"},
            params=params,
            timeout=sse_timeout,
        ) as res:
            if not res.is_success:
                await res.aread()
                raise _to_error(res, "events")
            async for frame in parse_sse(res, abort):
                yield SandboxEventAdapter.validate_python(json.loads(frame.data))


async def _sleep(seconds: float, abort: Optional[asyncio.Event]) -> None:
    """Sleep for `seconds`, waking early if `abort` is set."""
    if abort is None:
        await asyncio.sleep(seconds)
        return
    try:
        await asyncio.wait_for(abort.wait(), timeout=seconds)
    except asyncio.TimeoutError:
        pass

File: client/go/sandbox.go
package client

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"github.com/google/uuid"
)

// Sandbox is a handle to a provisioned sandbox. Obtain one via Client.GetOrCreateSandbox.
type Sandbox struct {
	// ID is the server-assigned unique identifier (UUID).
	ID string
	// Key is the caller-chosen key the sandbox was provisioned under.
	Key string

	apiURL string
	http   *http.Client
}

// ExecProcess is a handle to a running streaming exec. Obtain one via Sandbox.ExecStream.
type ExecProcess struct {
	// ID is the client-assigned exec identifier.
	ID string
	// Output receives stdout and stderr frames until the process exits.
	Output <-chan ExecOutput

	writeFn  func(ctx context.Context, data string) error
	exitCode <-chan int
	exitErr  <-chan error
}

// WriteStdin sends data to the process's stdin.
func (p *ExecProcess) WriteStdin(ctx context.Context, data string) error {
	return p.writeFn(ctx, data)
}

// Wait blocks until the process exits and returns its exit code.
func (p *ExecProcess) Wait() (int, error) {
	select {
	case code := <-p.exitCode:
		return code, nil
	case err := <-p.exitErr:
		return -1, err
	}
}

func newSandbox(ref SandboxRef, gatewayURL string, hc *http.Client) *Sandbox {
	base := strings.TrimRight(gatewayURL, "/")
	return &Sandbox{
		ID:     ref.ID,
		Key:    ref.Key,
		apiURL: fmt.Sprintf("%s/sandbox/%s", base, ref.ID),
		http:   hc,
	}
}

// keyed builds a per-sandbox API URL: <gateway>/sandbox/<id>/v1/<key><suffix>.
// suffix begins with "/" (e.g. "/config"), or is "" to address the sandbox
// resource itself (create/delete). Pod-level routes (/v1/ping) don't use this.
func (s *Sandbox) keyed(suffix string) string {
	return fmt.Sprintf("%s/v1/%s%s", s.apiURL, s.Key, suffix)
}

// fileURL addresses a file operation for the agent-visible absolute path,
// carried as trailing URL segments after /file (e.g. /file/workspace/data.csv).
// Each segment is escaped while the "/" separators are preserved, so a path with
// arbitrarily many segments round-trips intact.
func (s *Sandbox) fileURL(path string) string {
	var b strings.Builder
	for _, seg := range strings.Split(path, "/") {
		if seg == "" {
			continue
		}
		b.WriteByte('/')
		b.WriteString(url.PathEscape(seg))
	}
	return s.keyed("/file" + b.String())
}

// Shutdown tears the sandbox down via DELETE /v1/<key>, cancelling its lifecycle.
func (s *Sandbox) Shutdown(ctx context.Context) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodDelete, s.keyed(""), nil)
	if err != nil {
		return err
	}
	resp, err := s.http.Do(req)
	if err != nil {
		return fmt.Errorf("Shutdown: %w", err)
	}
	defer resp.Body.Close()
	if !isSuccess(resp.StatusCode) {
		return fmt.Errorf("Shutdown: %w", readAPIError(resp))
	}
	return nil
}

// ProxyURL returns the base URL for reaching port inside the sandbox.
// Append a path to form a full URL, e.g. sandbox.ProxyURL(8080) + "/health".
func (s *Sandbox) ProxyURL(port int) string {
	return s.keyed(fmt.Sprintf("/proxy/%d", port))
}

// Ping keeps the sandbox alive by resetting its TTL countdown.
func (s *Sandbox) Ping(ctx context.Context) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, s.keyed("/ping"), nil)
	if err != nil {
		return err
	}
	resp, err := s.http.Do(req)
	if err != nil {
		return fmt.Errorf("Ping: %w", err)
	}
	defer resp.Body.Close()
	if !isSuccess(resp.StatusCode) {
		return fmt.Errorf("Ping: %w", readAPIError(resp))
	}
	return nil
}

// GetConfig reads the current SandboxConfig.
func (s *Sandbox) GetConfig(ctx context.Context) (*SandboxConfig, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, s.keyed("/config"), nil)
	if err != nil {
		return nil, err
	}
	resp, err := s.http.Do(req)
	if err != nil {
		return nil, fmt.Errorf("GetConfig: %w", err)
	}
	defer resp.Body.Close()
	if !isSuccess(resp.StatusCode) {
		return nil, fmt.Errorf("GetConfig: %w", readAPIError(resp))
	}
	var cfg SandboxConfig
	if err := json.NewDecoder(resp.Body).Decode(&cfg); err != nil {
		return nil, fmt.Errorf("GetConfig: decode: %w", err)
	}
	return &cfg, nil
}

// GetInfo reads internal runtime information about the sandbox — currently the
// isolation mechanism in use, which sandboxd selects automatically from the
// image rather than from config.
func (s *Sandbox) GetInfo(ctx context.Context) (*SandboxInfo, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, s.keyed("/info"), nil)
	if err != nil {
		return nil, err
	}
	resp, err := s.http.Do(req)
	if err != nil {
		return nil, fmt.Errorf("GetInfo: %w", err)
	}
	defer resp.Body.Close()
	if !isSuccess(resp.StatusCode) {
		return nil, fmt.Errorf("GetInfo: %w", readAPIError(resp))
	}
	var info SandboxInfo
	if err := json.NewDecoder(resp.Body).Decode(&info); err != nil {
		return nil, fmt.Errorf("GetInfo: decode: %w", err)
	}
	return &info, nil
}

// Snapshot captures a snapshot of the running sandbox now, without stopping it.
// The request selects which parts to capture: VM (full microVM state, keyed for
// a later resume; a no-op on container isolation) and/or Files (the writable
// filesystem). Each part is reported independently in the result.
func (s *Sandbox) Snapshot(ctx context.Context, req Snapshot) (*SnapshotResult, error) {
	body, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("Snapshot: encode: %w", err)
	}
	hreq, err := http.NewRequestWithContext(ctx, http.MethodPost, s.keyed("/snapshot"), bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	hreq.Header.Set("Content-Type", "application/json")

	resp, err := s.http.Do(hreq)
	if err != nil {
		return nil, fmt.Errorf("Snapshot: %w", err)
	}
	defer resp.Body.Close()
	if !isSuccess(resp.StatusCode) {
		return nil, fmt.Errorf("Snapshot: %w", readAPIError(resp))
	}
	var result SnapshotResult
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("Snapshot: decode: %w", err)
	}
	return &result, nil
}

// ApplyConfig applies a desired SandboxConfig. The result's Applied field
// reports whether the change was committed or rolled back.
func (s *Sandbox) ApplyConfig(ctx context.Context, config SandboxConfig) (*ApplyResult, error) {
	body, err := json.Marshal(config)
	if err != nil {
		return nil, fmt.Errorf("ApplyConfig: encode: %w", err)
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPut, s.keyed("/config"), bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := s.http.Do(req)
	if err != nil {
		return nil, fmt.Errorf("ApplyConfig: %w", err)
	}
	defer resp.Body.Close()
	if !isSuccess(resp.StatusCode) {
		return nil, fmt.Errorf("ApplyConfig: %w", readAPIError(resp))
	}
	var result ApplyResult
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("ApplyConfig: decode: %w", err)
	}
	return &result, nil
}

// GetPorts returns the TCP ports the sandbox currently exposes.
// Each port is reachable via ProxyURL.
func (s *Sandbox) GetPorts(ctx context.Context) ([]int, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, s.keyed("/ports"), nil)
	if err != nil {
		return nil, err
	}
	resp, err := s.http.Do(req)
	if err != nil {
		return nil, fmt.Errorf("GetPorts: %w", err)
	}
	defer resp.Body.Close()
	if !isSuccess(resp.StatusCode) {
		return nil, fmt.Errorf("GetPorts: %w", readAPIError(resp))
	}
	var ports []int
	if err := json.NewDecoder(resp.Body).Decode(&ports); err != nil {
		return nil, fmt.Errorf("GetPorts: decode: %w", err)
	}
	return ports, nil
}

// Exec runs a command inside the sandbox and returns the buffered result once
// the process exits. Use ExecStream when you need output incrementally.
func (s *Sandbox) Exec(ctx context.Context, req ExecRequest) (*ExecResult, error) {
	body, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("Exec: encode: %w", err)
	}
	hreq, err := http.NewRequestWithContext(ctx, http.MethodPost, s.keyed("/exec"), bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	hreq.Header.Set("Content-Type", "application/json")

	resp, err := s.http.Do(hreq)
	if err != nil {
		return nil, fmt.Errorf("Exec: %w", err)
	}
	defer resp.Body.Close()
	if !isSuccess(resp.StatusCode) {
		return nil, fmt.Errorf("Exec: %w", readAPIError(resp))
	}
	var result ExecResult
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("Exec: decode: %w", err)
	}
	return &result, nil
}

// ExecStream starts a streaming exec and returns an ExecProcess handle.
// Drain Output to receive stdout/stderr frames, call WriteStdin to write
// to the process, and call Wait to block until exit. Pass an empty command
// in req to attach to the sandbox entrypoint's TTY (requires tty:true).
func (s *Sandbox) ExecStream(ctx context.Context, req ExecStreamRequest) (*ExecProcess, error) {
	execID := uuid.New().String()

	body, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("ExecStream: encode: %w", err)
	}

	streamURL := s.keyed("/exec-stream/"+execID)
	stdinURL := s.keyed("/exec-stream/"+execID+"/stdin")

	hreq, err := http.NewRequestWithContext(ctx, http.MethodPost, streamURL, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	hreq.Header.Set("Content-Type", "application/json")
	hreq.Header.Set("Accept", "text/event-stream")

	resp, err := s.http.Do(hreq)
	if err != nil {
		return nil, fmt.Errorf("ExecStream: %w", err)
	}
	if !isSuccess(resp.StatusCode) {
		defer resp.Body.Close()
		return nil, fmt.Errorf("ExecStream: %w", readAPIError(resp))
	}

	outputCh := make(chan ExecOutput, 32)
	exitCodeCh := make(chan int, 1)
	exitErrCh := make(chan error, 1)

	go func() {
		defer resp.Body.Close()
		defer close(outputCh)
		for frame := range readSSE(resp.Body) {
			var event struct {
				Type string `json:"type"`
				Text string `json:"text"`
				Code int    `json:"code"`
			}
			if err := json.Unmarshal([]byte(frame.data), &event); err != nil {
				continue
			}
			switch event.Type {
			case "stdout":
				outputCh <- ExecOutput{Stdout: event.Text}
			case "stderr":
				outputCh <- ExecOutput{Stderr: event.Text}
			case "exit":
				exitCodeCh <- event.Code
				return
			}
		}
		exitErrCh <- fmt.Errorf("ExecStream: stream closed without exit frame")
	}()

	hc := s.http
	writeFn := func(ctx context.Context, data string) error {
		body, _ := json.Marshal(map[string]string{"data": data})
		req, err := http.NewRequestWithContext(ctx, http.MethodPost, stdinURL, bytes.NewReader(body))
		if err != nil {
			return err
		}
		req.Header.Set("Content-Type", "application/json")
		resp, err := hc.Do(req)
		if err != nil {
			return fmt.Errorf("WriteStdin: %w", err)
		}
		defer resp.Body.Close()
		if resp.StatusCode != http.StatusNoContent {
			return fmt.Errorf("WriteStdin: %w", readAPIError(resp))
		}
		return nil
	}

	return &ExecProcess{
		ID:       execID,
		Output:   outputCh,
		writeFn:  writeFn,
		exitCode: exitCodeCh,
		exitErr:  exitErrCh,
	}, nil
}

// ListDirectory returns the immediate children of path inside the sandbox.
// path is the agent-visible absolute path (e.g. /workspace).
func (s *Sandbox) ListDirectory(ctx context.Context, path string) ([]DirEntry, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, s.keyed("/directories"), nil)
	if err != nil {
		return nil, err
	}
	q := req.URL.Query()
	q.Set("path", path)
	req.URL.RawQuery = q.Encode()

	resp, err := s.http.Do(req)
	if err != nil {
		return nil, fmt.Errorf("ListDirectory: %w", err)
	}
	defer resp.Body.Close()
	if !isSuccess(resp.StatusCode) {
		return nil, fmt.Errorf("ListDirectory: %w", readAPIError(resp))
	}
	var result struct {
		Entries []DirEntry `json:"entries"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("ListDirectory: decode: %w", err)
	}
	return result.Entries, nil
}

// ReadFile returns the raw bytes of the file at path.
// path is the agent-visible absolute path (e.g. /workspace/data.csv).
func (s *Sandbox) ReadFile(ctx context.Context, path string) ([]byte, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, s.fileURL(path), nil)
	if err != nil {
		return nil, err
	}

	resp, err := s.http.Do(req)
	if err != nil {
		return nil, fmt.Errorf("ReadFile: %w", err)
	}
	defer resp.Body.Close()
	if !isSuccess(resp.StatusCode) {
		return nil, fmt.Errorf("ReadFile: %w", readAPIError(resp))
	}
	return io.ReadAll(resp.Body)
}

// WriteFile writes content to path inside the sandbox.
// path is the agent-visible absolute path (e.g. /workspace/data.csv) and must
// resolve beneath one of the configured fs[].mount paths.
func (s *Sandbox) WriteFile(ctx context.Context, path string, content []byte) (*UploadResult, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, s.fileURL(path), bytes.NewReader(content))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/octet-stream")

	resp, err := s.http.Do(req)
	if err != nil {
		return nil, fmt.Errorf("WriteFile: %w", err)
	}
	defer resp.Body.Close()
	if !isSuccess(resp.StatusCode) {
		return nil, fmt.Errorf("WriteFile: %w", readAPIError(resp))
	}
	var result UploadResult
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("WriteFile: decode: %w", err)
	}
	return &result, nil
}

// DeleteFile removes the file or empty directory at path inside the sandbox.
// path is the agent-visible absolute path (e.g. /workspace/data.csv).
func (s *Sandbox) DeleteFile(ctx context.Context, path string) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodDelete, s.fileURL(path), nil)
	if err != nil {
		return err
	}

	resp, err := s.http.Do(req)
	if err != nil {
		return fmt.Errorf("DeleteFile: %w", err)
	}
	defer resp.Body.Close()
	if !isSuccess(resp.StatusCode) {
		return fmt.Errorf("DeleteFile: %w", readAPIError(resp))
	}
	return nil
}

// WatchEvents streams the sandbox's activity events (egress, filesystem, exec,
// stdio, resource usage) as they happen. It auto-resumes across transient
// disconnects up to 3 times with exponential backoff. Pass lastEventID = -1 to
// start from the next new event, or a non-negative value to resume after a
// specific event.
//
// The returned channel is closed when ctx is done or retries are exhausted.
// Any terminal error is sent on errc.
func (s *Sandbox) WatchEvents(ctx context.Context, lastEventID int) (<-chan SandboxEvent, <-chan error) {
	events := make(chan SandboxEvent)
	errc := make(chan error, 1)

	go func() {
		defer close(events)
		const maxRetries = 3
		backoff := 200 * time.Millisecond
		retries := 0

		for {
			if ctx.Err() != nil {
				return
			}
			if retries > maxRetries {
				errc <- fmt.Errorf("WatchEvents: max reconnect retries exceeded")
				return
			}

			lastID, err := s.streamEvents(ctx, lastEventID, events)
			lastEventID = lastID
			if err == nil || ctx.Err() != nil {
				return
			}

			retries++
			select {
			case <-time.After(backoff):
				backoff = min(backoff*2, 30*time.Second)
			case <-ctx.Done():
				return
			}
		}
	}()

	return events, errc
}

func (s *Sandbox) streamEvents(ctx context.Context, lastEventID int, out chan<- SandboxEvent) (int, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, s.keyed("/events"), nil)
	if err != nil {
		return lastEventID, err
	}
	req.Header.Set("Accept", "text/event-stream")
	if lastEventID >= 0 {
		q := req.URL.Query()
		q.Set("lastEventId", strconv.Itoa(lastEventID))
		req.URL.RawQuery = q.Encode()
	}

	resp, err := s.http.Do(req)
	if err != nil {
		return lastEventID, err
	}
	defer resp.Body.Close()

	if !isSuccess(resp.StatusCode) {
		return lastEventID, fmt.Errorf("status %d", resp.StatusCode)
	}

	for frame := range readSSE(resp.Body) {
		var evt SandboxEvent
		if err := json.Unmarshal([]byte(frame.data), &evt); err != nil {
			continue
		}
		lastEventID = evt.ID
		select {
		case out <- evt:
		case <-ctx.Done():
			return lastEventID, nil
		}
	}
	return lastEventID, nil
}

func isSuccess(code int) bool {
	return code >= 200 && code < 300
}
